<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <title>Special Topics: Gaussian Processes</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="https://unpkg.com/reveal.js@3.9.2/css/reveal.css">
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="https://unpkg.com/reveal.js@3.9.2/css/theme/black.css" id="theme">
  <link rel="stylesheet" href="https://inverseprobability.com/assets/css/talks.css"/>
  <!-- Printing and PDF exports -->
  <script>
    var link = document.createElement( 'link' );
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = window.location.search.match( /print-pdf/gi ) ? 'https://unpkg.com/reveal.js@3.9.2/css/print/pdf.css' : 'https://unpkg.com/reveal.js@3.9.2/css/print/paper.css';
    document.getElementsByTagName( 'head' )[0].appendChild( link );
  </script>
  <!--[if lt IE 9]>
  <script src="https://unpkg.com/reveal.js@3.9.2/lib/js/html5shiv.js"></script>
  <![endif]-->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_SVG" type="text/javascript"></script>
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
    TeX: {
         extensions: ["color.js"]
      }
    });
  </script>
  <script src="../assets/js/figure-animate.js"></script>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Special Topics: Gaussian Processes</h1>
</section>

<section class="slide level2">

<!-- Do not edit this file locally. -->
<!---->
<!-- Do not edit this file locally. -->
<!-- The last names to be defined. Should be defined entirely in terms of macros from above-->
<!--setupplotcode{import seaborn as sns
sns.set_style('darkgrid')
sns.set_context('paper')
sns.set_palette('colorblind')}-->
</section>
<section id="review" class="slide level2">
<h2>Review</h2>
<ul>
<li>Last week: Logistic Regression and Generalised Linear Models</li>
<li>Introduced link functions and different transformations.</li>
<li>Showed examples in classification and mentioned possibilities for
disease rate models.</li>
<li>This week:
<ul>
<li>Gaussian Processes: non parametric Bayesian modelling</li>
</ul></li>
</ul>
</section>
<section id="generalized-linear-models" class="slide level2">
<h2>Generalized Linear Models</h2>
<ul>
<li>Logistic regression is a <em>generalized linear model</em></li>
<li>Prediction function that is linear in parameters <span
class="math display">\[
\log \frac{p(\mathbf{ x})}{1-p(\mathbf{ x})} = f(\mathbf{ x}; \mathbf{
w})
\]</span></li>
<li>Where <span class="math inline">\(f(\cdot)\)</span> is linear in
parameters.</li>
</ul>
</section>
<section id="basis-functions" class="slide level2">
<h2>Basis Functions</h2>
<ul>
<li><p>Linear models have the form <span class="math display">\[
f(\mathbf{ x}) =
\mathbf{ w}^\top \boldsymbol{ \phi}(\mathbf{ x}).
\]</span></p></li>
<li><p>Gaussian processes are related to generalized linear
models.</p></li>
</ul>
</section>
<section id="gaussian-processes" class="slide level2">
<h2>Gaussian Processes</h2>
<ul>
<li>Basis function models give non-linear predictions.</li>
<li>Need to choose number and location of basis functions.</li>
<li>Gaussian processes is a general framework (basis functions special
case)</li>
<li>Within the framework you can consider models with infinite basis
functions.</li>
</ul>
<p><span class="math display">\[
p(\mathbf{ y}|\mathbf{X}, \mathbf{ w}) =
\prod_{i=1}^{n} p(y_i | \mathbf{ x}_i, \mathbf{ w})
\]</span></p>
<p><span class="math display">\[
\mathbf{ y}|\mathbf{X}\sim
\mathcal{N}\left(\mathbf{m}(\mathbf{X}),\mathbf{K}(\mathbf{X})\right),
\]</span></p>
</section>
<section id="sampling-a-function" class="slide level2">
<h2>Sampling a Function</h2>
<p><strong>Multi-variate Gaussians</strong></p>
<ul>
<li>We will consider a Gaussian with a particular structure of
covariance matrix.</li>
<li>Generate a single sample from this 25 dimensional Gaussian density,
<span class="math display">\[
\mathbf{ f}=\left[f_{1},f_{2}\dots f_{25}\right].
\]</span></li>
<li>We will plot these points against their index.</li>
</ul>
</section>
<section id="gaussian-distribution-sample" class="slide level2">
<h2>Gaussian Distribution Sample</h2>
<script>
showDivs(0, 'two_point_sample');
</script>
<p><small></small>
<input id="range-two_point_sample" type="range" min="0" max="8" value="0" onchange="setDivs('two_point_sample')" oninput="setDivs('two_point_sample')">
<button onclick="plusDivs(-1, 'two_point_sample')">❮</button>
<button onclick="plusDivs(1, 'two_point_sample')">❯</button></p>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample000.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample001.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample002.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample003.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample004.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample005.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample006.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample007.svg" width style=" ">
</object>
</div>
<div class="two_point_sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample008.svg" width style=" ">
</object>
</div>
</section>
<section id="sampling-a-function-from-a-gaussian" class="slide level2">
<h2>Sampling a Function from a Gaussian</h2>
<script>
showDivs(9, 'two-point-sample');
</script>
<p><small></small>
<input id="range-two-point-sample" type="range" min="9" max="12" value="9" onchange="setDivs('two-point-sample')" oninput="setDivs('two-point-sample')">
<button onclick="plusDivs(-1, 'two-point-sample')">❮</button>
<button onclick="plusDivs(1, 'two-point-sample')">❯</button></p>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample000.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample001.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample002.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample003.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample004.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample005.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample006.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample007.svg" width="80%" style=" ">
</object>
</div>
<div class="two-point-sample" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample008.svg" width="80%" style=" ">
</object>
</div>
</section>
<section id="joint-density-of-f_1-and-f_2" class="slide level2">
<h2>Joint Density of <span class="math inline">\(f_1\)</span> and <span
class="math inline">\(f_2\)</span></h2>
</section>
<section id="prediction-of-f_2-from-f_1" class="slide level2">
<h2>Prediction of <span class="math inline">\(f_{2}\)</span> from <span
class="math inline">\(f_{1}\)</span></h2>
<script>
showDivs(9, 'two_point_sample2');
</script>
<p><small></small>
<input id="range-two_point_sample2" type="range" min="9" max="12" value="9" onchange="setDivs('two_point_sample2')" oninput="setDivs('two_point_sample2')">
<button onclick="plusDivs(-1, 'two_point_sample2')">❮</button>
<button onclick="plusDivs(1, 'two_point_sample2')">❯</button></p>
<div class="two_point_sample2" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample009.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample2" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample010.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample2" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample011.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample2" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample012.svg" width="80%" style=" ">
</object>
</div>
</section>
<section id="uluru" class="slide level2">
<h2>Uluru</h2>
<div class="figure">
<div id="uluru-as-probability-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/799px-Uluru_Panorama.jpg" width="" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Uluru, the sacred rock in Australia. If we think of it as a probability
density, viewing it from this side gives us one <em>marginal</em> from
the density. Figuratively speaking, slicing through the rock would give
a conditional density.
</aside>
</section>
<section id="prediction-with-correlated-gaussians" class="slide level2">
<h2>Prediction with Correlated Gaussians</h2>
<ul>
<li>Prediction of <span class="math inline">\(f_2\)</span> from <span
class="math inline">\(f_1\)</span> requires <em>conditional
density</em>.</li>
<li>Conditional density is <em>also</em> Gaussian. <span
class="math display">\[
p(f_2|f_1) = \mathcal{N}\left(f_2|\frac{k_{1, 2}}{k_{1, 1}}f_1, k_{2, 2}
- \frac{k_{1,2}^2}{k_{1,1}}\right)
\]</span> where covariance of joint density is given by <span
class="math display">\[
\mathbf{K}= \begin{bmatrix} k_{1, 1} &amp; k_{1, 2}\\ k_{2, 1} &amp;
k_{2, 2}.\end{bmatrix}
\]</span></li>
</ul>
</section>
<section id="joint-density-of-f_1-and-f_8" class="slide level2">
<h2>Joint Density of <span class="math inline">\(f_1\)</span> and <span
class="math inline">\(f_8\)</span></h2>
</section>
<section id="prediction-of-f_8-from-f_1" class="slide level2">
<h2>Prediction of <span class="math inline">\(f_{8}\)</span> from <span
class="math inline">\(f_{1}\)</span></h2>
<script>
showDivs(13, 'two_point_sample3');
</script>
<p><small></small>
<input id="range-two_point_sample3" type="range" min="13" max="17" value="13" onchange="setDivs('two_point_sample3')" oninput="setDivs('two_point_sample3')">
<button onclick="plusDivs(-1, 'two_point_sample3')">❮</button>
<button onclick="plusDivs(1, 'two_point_sample3')">❯</button></p>
<div class="two_point_sample3" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample013.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample3" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample014.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample3" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample015.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample3" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample016.svg" width="80%" style=" ">
</object>
</div>
<div class="two_point_sample3" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/two_point_sample017.svg" width="80%" style=" ">
</object>
</div>
</section>
<section id="details" class="slide level2">
<h2>Details</h2>
<ul>
<li>The single contour of the Gaussian density represents the
<font color="yellow">joint distribution, <span
class="math inline">\(p(f_1, f_8)\)</span></font></li>
</ul>
<div class="fragment">
<ul>
<li>We observe a value for <font color="magenta"><span
class="math inline">\(f_1=-?\)</span></font></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Conditional density: <font color="cyan"><span
class="math inline">\(p(f_8|f_1=?)\)</span></font>.</li>
</ul>
</div>
</section>
<section id="prediction-with-correlated-gaussians-1"
class="slide level2">
<h2>Prediction with Correlated Gaussians</h2>
<ul>
<li><p>Prediction of <span class="math inline">\(\mathbf{ f}_*\)</span>
from <span class="math inline">\(\mathbf{ f}\)</span> requires
multivariate <em>conditional density</em>.</p></li>
<li><p>Multivariate conditional density is <em>also</em> Gaussian.
<large> <span class="math display">\[
p(\mathbf{ f}_*|\mathbf{ f}) = {\mathcal{N}\left(\mathbf{
f}_*|\mathbf{K}_{*,\mathbf{ f}}\mathbf{K}_{\mathbf{ f},\mathbf{
f}}^{-1}\mathbf{ f},\mathbf{K}_{*,*}-\mathbf{K}_{*,\mathbf{ f}}
\mathbf{K}_{\mathbf{ f},\mathbf{ f}}^{-1}\mathbf{K}_{\mathbf{
f},*}\right)}
\]</span> </large></p></li>
<li><p>Here covariance of joint density is given by <span
class="math display">\[
\mathbf{K}= \begin{bmatrix} \mathbf{K}_{\mathbf{ f}, \mathbf{ f}} &amp;
\mathbf{K}_{*, \mathbf{ f}}\\ \mathbf{K}_{\mathbf{ f}, *} &amp;
\mathbf{K}_{*, *}\end{bmatrix}
\]</span></p></li>
</ul>
</section>
<section id="prediction-with-correlated-gaussians-2"
class="slide level2">
<h2>Prediction with Correlated Gaussians</h2>
<ul>
<li><p>Prediction of <span class="math inline">\(\mathbf{ f}_*\)</span>
from <span class="math inline">\(\mathbf{ f}\)</span> requires
multivariate <em>conditional density</em>.</p></li>
<li><p>Multivariate conditional density is <em>also</em> Gaussian.
<large> <span class="math display">\[
p(\mathbf{ f}_*|\mathbf{ f}) = {\mathcal{N}\left(\mathbf{
f}_*|\boldsymbol{ \mu},\boldsymbol{ \Sigma}\right)}
\]</span> <span class="math display">\[
\boldsymbol{ \mu}= \mathbf{K}_{*,\mathbf{ f}}\mathbf{K}_{\mathbf{
f},\mathbf{ f}}^{-1}\mathbf{ f}
\]</span> <span class="math display">\[
\boldsymbol{ \Sigma}= \mathbf{K}_{*,*}-\mathbf{K}_{*,\mathbf{ f}}
\mathbf{K}_{\mathbf{ f},\mathbf{ f}}^{-1}\mathbf{K}_{\mathbf{ f},*}
\]</span> </large></p></li>
<li><p>Here covariance of joint density is given by <span
class="math display">\[
\mathbf{K}= \begin{bmatrix} \mathbf{K}_{\mathbf{ f}, \mathbf{ f}} &amp;
\mathbf{K}_{*, \mathbf{ f}}\\ \mathbf{K}_{\mathbf{ f}, *} &amp;
\mathbf{K}_{*, *}\end{bmatrix}
\]</span></p></li>
</ul>
</section>
<section id="marginal-likelihood" class="slide level2">
<h2>Marginal Likelihood</h2>
</section>
<section id="sampling-from-the-prior" class="slide level2">
<h2>Sampling from the Prior</h2>
</section>
<section id="weight-space-view" class="slide level2">
<h2>Weight Space View</h2>
<p><span class="math display">\[
\mathbf{ w}\sim \mathcal{N}\left(\mathbf{0},\alpha\mathbf{I}\right),
\]</span></p>
<p><span class="math display">\[
f(\mathbf{ x}) = \mathbf{ w}^\top \boldsymbol{ \phi}(\mathbf{ x}).
\]</span></p>
</section>
<section id="function-space-view" class="slide level2">
<h2>Function Space View</h2>
<p><span class="math display">\[
\mathbf{ w}\sim \mathcal{N}\left(\mathbf{0},\alpha \mathbf{I}\right)
\]</span></p>
<p><span class="math display">\[
\boldsymbol{ \Phi}= \begin{bmatrix}\boldsymbol{ \phi}(\mathbf{ x}_1) \\
\vdots \\
\boldsymbol{ \phi}(\mathbf{ x}_n)\end{bmatrix}
\]</span></p>
<p><span class="math display">\[
\mathbf{ f}= \begin{bmatrix} f_1
\\ \vdots f_n\end{bmatrix}
\]</span></p>
<p><span class="math display">\[
\mathbf{ f}= \boldsymbol{ \Phi}\mathbf{ w}.
\]</span></p>
</section>
<section id="section" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\mathbf{ f}\sim \mathcal{N}\left(\mathbf{0},\alpha \boldsymbol{
\Phi}\boldsymbol{ \Phi}^\top\right).
\]</span></p>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\mathbf{K}= \alpha
\boldsymbol{ \Phi}\boldsymbol{ \Phi}^\top.
\]</span></p>
<div class="sourceCode" id="cb1"><pre
class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>K <span class="op">=</span> alpha<span class="op">*</span>Phi_pred<span class="op">@</span>Phi_pred.T</span></code></pre></div>
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<div class="sourceCode" id="cb2"><pre
class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>K <span class="op">=</span> alpha<span class="op">*</span>Phi_pred<span class="op">@</span>Phi_pred.T</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>f_sample <span class="op">=</span> np.random.multivariate_normal(mean<span class="op">=</span>np.zeros(x_pred.size), cov<span class="op">=</span>K)</span></code></pre></div>
<div class="figure">
<div id="gp-sample-basis-function-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//kern/gp-sample-basis-function.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Samples directly from the covariance function implied by the basis
function based covariance, <span class="math inline">\(\alpha
\boldsymbol{ \Phi}\boldsymbol{ \Phi}^\top\)</span>.
</aside>
</section>
<section id="section-3" class="slide level2">
<h2></h2>
</section>
<section id="section-4" class="slide level2">
<h2></h2>
<div class="figure">
<div id="basis-covariance-function-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//kern/basis-covariance-function.svg" width="60%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Covariance of the function implied by the basis set <span
class="math inline">\(\alpha\boldsymbol{ \Phi}\boldsymbol{
\Phi}^\top\)</span>.
</aside>
</section>
<section id="section-5" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\mathbf{ y}= \mathbf{ f}+ \boldsymbol{\epsilon}
\]</span></p>
<p><span class="math display">\[
\epsilon \sim \mathcal{N}\left(\mathbf{0},\sigma^2\mathbf{I}\right).
\]</span></p>
<p><span class="math display">\[
\mathbf{ y}\sim \mathcal{N}\left(\mathbf{0},\boldsymbol{
\Phi}\boldsymbol{ \Phi}^\top +\sigma^2\mathbf{I}\right).
\]</span></p>
</section>
<section id="section-6" class="slide level2">
<h2></h2>
<div class="figure">
<div id="gp-sample-basis-functions-plus-noise-figure"
class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//kern/gp-sample-basis-function-plus-noise.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Samples directly from the covariance function implied by the noise
corrupted basis function based covariance, <span
class="math inline">\(\alpha \boldsymbol{ \Phi}\boldsymbol{ \Phi}^\top +
\sigma^2 \mathbf{I}\)</span>.
</aside>
<div class="figure">
<div id="gp-sample-basis-functions-plus-large-noise-figure"
class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//kern/gp-sample-basis-function-plus-large-noise.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Samples directly from the covariance function implied by the noise
corrupted basis function based covariance, <span
class="math inline">\(\alpha \boldsymbol{ \Phi}\boldsymbol{ \Phi}^\top +
\mathbf{I}\)</span>.
</aside>
</section>
<section id="non-degenerate-gaussian-processes" class="slide level2">
<h2>Non-degenerate Gaussian Processes</h2>
<ul>
<li>This process is <em>degenerate</em>.</li>
<li>Covariance function is of rank at most <span
class="math inline">\(h\)</span>.</li>
<li>As <span class="math inline">\(n\rightarrow \infty\)</span>,
covariance matrix is not full rank.</li>
<li>Leading to <span class="math inline">\(\det{\mathbf{K}} =
0\)</span></li>
</ul>
</section>
<section id="infinite-networks" class="slide level2">
<h2>Infinite Networks</h2>
<div class="centered" style="">
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip0">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
Radford Neal
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/radford-neal.jpg" clip-path="url(#clip0)"/>
</svg>
</div>
<ul>
<li>In ML Radford Neal <span class="citation"
data-cites="Neal:bayesian94">(Neal, 1994)</span> asked “what would
happen if you took <span class="math inline">\(h\rightarrow
\infty\)</span>?”</li>
</ul>
<div class="figure">
<div id="neal-infinite-priors-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//neal-infinite-priors.png" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Page 37 of <a
href="http://www.cs.toronto.edu/~radford/ftp/thesis.pdf">Radford Neal’s
1994 thesis</a>
</aside>
</section>
<section id="roughly-speaking" class="slide level2">
<h2>Roughly Speaking</h2>
<ul>
<li>Instead of <span class="math display">\[
\begin{align*}
k_f\left(\mathbf{ x}_i, \mathbf{ x}_j\right) &amp; = \alpha \boldsymbol{
\phi}\left(\mathbf{W}_1, \mathbf{ x}_i\right)^\top \boldsymbol{
\phi}\left(\mathbf{W}_1, \mathbf{ x}_j\right)\\
&amp; = \alpha \sum_k \phi\left(\mathbf{ w}^{(1)}_k, \mathbf{
x}_i\right) \phi\left(\mathbf{ w}^{(1)}_k, \mathbf{ x}_j\right)
\end{align*}
\]</span></li>
<li>Sample infinitely many from a prior density, <span
class="math inline">\(p(\mathbf{ w}^{(1)})\)</span>, <span
class="math display">\[
k_f\left(\mathbf{ x}_i, \mathbf{ x}_j\right) = \alpha \int
\phi\left(\mathbf{ w}^{(1)}, \mathbf{ x}_i\right) \phi\left(\mathbf{
w}^{(1)}, \mathbf{ x}_j\right) p(\mathbf{ w}^{(1)}) \text{d}\mathbf{
w}^{(1)}
\]</span></li>
<li>Also applies for non-Gaussian <span class="math inline">\(p(\mathbf{
w}^{(1)})\)</span> because of the <em>central limit theorem</em>.</li>
</ul>
</section>
<section id="simple-probabilistic-program" class="slide level2">
<h2>Simple Probabilistic Program</h2>
<ul>
<li><p>If <span class="math display">\[
\begin{align*}
\mathbf{ w}^{(1)} &amp; \sim p(\cdot)\\ \phi_i &amp; =
\phi\left(\mathbf{ w}^{(1)}, \mathbf{ x}_i\right),
\end{align*}
\]</span> has finite variance.</p></li>
<li><p>Then taking number of hidden units to infinity, is also a
Gaussian process.</p></li>
</ul>
</section>
<section id="further-reading" class="slide level2">
<h2>Further Reading</h2>
<ul>
<li><p>Chapter 2 of Neal’s thesis <span class="citation"
data-cites="Neal:bayesian94">(Neal, 1994)</span></p></li>
<li><p>Rest of Neal’s thesis. <span class="citation"
data-cites="Neal:bayesian94">(Neal, 1994)</span></p></li>
<li><p>David MacKay’s PhD thesis <span class="citation"
data-cites="MacKay:bayesian92">(MacKay, 1992)</span></p></li>
</ul>
</section>
<section id="gaussian-process" class="slide level2">
<h2>Gaussian Process</h2>
<p><span class="math display">\[
k(\mathbf{ x}, \mathbf{ x}^\prime) = \alpha \exp\left( -\frac{\left\Vert
\mathbf{ x}-\mathbf{ x}^\prime\right\Vert^2}{2\ell^2}\right),
\]</span></p>
<p><span class="math display">\[
\left\Vert\mathbf{ x}- \mathbf{ x}^\prime\right\Vert^2 = (\mathbf{ x}-
\mathbf{ x}^\prime)^\top (\mathbf{ x}- \mathbf{ x}^\prime)
\]</span></p>
</section>
<section id="section-7" class="slide level2">
<h2></h2>
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp_rejection_sample001.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<aside class="notes">
Here we’re showing 20 samples taken from the prior over functions
defined by our covarariance
</aside>
</section>
<section id="section-8" class="slide level2">
<h2></h2>
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp_rejection_sample002.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<aside class="notes">
We can sample many such functions, in this slide there are now 1000 in
total. This is a sample from our prior over functions.
</aside>
</section>
<section id="section-9" class="slide level2">
<h2></h2>
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp_rejection_sample003.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<aside class="notes">
Now we observe data. Here there are three data points. Conceptually in
Bayesian inference we discard all samples that are distant from the
data.
</aside>
</section>
<section id="section-10" class="slide level2">
<h2></h2>
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp_rejection_sample004.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<aside class="notes">
Throwing away such samples we are left with our posterior. This is the
collection of samples from the prior that are consistent with the data.
</aside>
</section>
<section id="section-11" class="slide level2">
<h2></h2>
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp_rejection_sample005.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<aside class="notes">
The elegance of the Gaussian process is that this result can be computed
analytically using linear algebra.
</aside>
</section>
<section id="gaussian-process-1" class="slide level2">
<h2>Gaussian Process</h2>
<p><span class="math display">\[
p(\mathbf{ y}|\mathbf{X}) =
\frac{1}{(2\pi)^{\frac{n}{2}}|\mathbf{K}|^{\frac{1}{2}}}
\exp\left(-\frac{1}{2}\mathbf{ y}^\top \left(\mathbf{K}+\sigma^2
\mathbf{I}\right)^{-1}\mathbf{ y}\right)
\]</span></p>
<p><span class="math display">\[
E(\boldsymbol{\theta}) = \frac{1}{2} \log |\mathbf{K}|
+ \frac{1}{2} \mathbf{ y}^\top \left(\mathbf{K}+
\sigma^2\mathbf{I}\right)^{-1}\mathbf{ y}
\]</span></p>
</section>
<section id="making-predictions" class="slide level2">
<h2>Making Predictions</h2>
<p><span class="math display">\[
\begin{bmatrix}\mathbf{ f}\\ \mathbf{ f}^*\end{bmatrix} \sim
\mathcal{N}\left(\mathbf{0},\begin{bmatrix} \mathbf{K}&amp;
\mathbf{K}_\ast \\
\mathbf{K}_\ast^\top &amp; \mathbf{K}_{\ast,\ast}\end{bmatrix}\right)
\]</span></p>
<p><span class="math display">\[
\begin{bmatrix} \mathbf{K}&amp; \mathbf{K}_\ast \\ \mathbf{K}_\ast^\top
&amp; \mathbf{K}_{\ast,\ast}\end{bmatrix}
\]</span></p>
<div class="figure">
<div id="block-predictive-covariance-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/block-predictive-covariance.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Different blocks of the covariance function. The upper left block is the
covariance of the training data with itself, <span
class="math inline">\(\mathbf{K}\)</span>. The top right is the cross
covariance between training data (rows) and prediction locations
(columns). The lower left is the same matrix transposed. The bottom
right is the covariance matrix of the test data with itself.
</aside>
</section>
<section id="the-importance-of-the-covariance-function"
class="slide level2">
<h2>The Importance of the Covariance Function</h2>
<p><span class="math display">\[
\boldsymbol{ \mu}_f= \mathbf{A}^\top \mathbf{ y},
\]</span></p>
<div class="figure">
<div id="intro-to-gps-figure" class="figure-frame">
<iframe width="600" height="450" src="https://www.youtube.com/embed/ewJ3AxKclOg?start=" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen>
</iframe>
</div>
</div>
<aside class="notes">
Introduction to Gaussian processes given by Neil Lawrence at the 2014
Gaussian process Winter School at the University of Sheffield.
</aside>
</section>
<section id="improving-the-numerics" class="slide level2">
<h2>Improving the Numerics</h2>
<p>In practice we shouldn’t be using matrix inverse directly to solve
the GP system. One more stable way is to compute the <em>Cholesky
decomposition</em> of the kernel matrix. The log determinant of the
covariance can also be derived from the Cholesky decomposition.</p>
</section>
<section id="capacity-control" class="slide level2">
<h2>Capacity Control</h2>
</section>
<section id="gradients-of-the-likelihood" class="slide level2">
<h2>Gradients of the Likelihood</h2>
</section>
<section id="overall-process-scale" class="slide level2">
<h2>Overall Process Scale</h2>
</section>
<section id="capacity-control-and-data-fit" class="slide level2">
<h2>Capacity Control and Data Fit</h2>
</section>
<section id="learning-covariance-parameters" class="slide level2">
<h2>Learning Covariance Parameters</h2>
<p>Can we determine covariance parameters from the data?</p>
</section>
<section id="section-12" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\mathcal{N}\left(\mathbf{
y}|\mathbf{0},\mathbf{K}\right)=\frac{1}{(2\pi)^\frac{n}{2}{\det{\mathbf{K}}^{\frac{1}{2}}}}{\exp\left(-\frac{\mathbf{
y}^{\top}\mathbf{K}^{-1}\mathbf{ y}}{2}\right)}
\]</span></p>
</section>
<section id="section-13" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\begin{aligned}
    \mathcal{N}\left(\mathbf{
y}|\mathbf{0},\mathbf{K}\right)=\frac{1}{(2\pi)^\frac{n}{2}\color{yellow}{\det{\mathbf{K}}^{\frac{1}{2}}}}\color{cyan}{\exp\left(-\frac{\mathbf{
y}^{\top}\mathbf{K}^{-1}\mathbf{ y}}{2}\right)}
\end{aligned}
\]</span></p>
</section>
<section id="section-14" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\begin{aligned}
    \log \mathcal{N}\left(\mathbf{
y}|\mathbf{0},\mathbf{K}\right)=&amp;\color{yellow}{-\frac{1}{2}\log\det{\mathbf{K}}}\color{cyan}{-\frac{\mathbf{
y}^{\top}\mathbf{K}^{-1}\mathbf{ y}}{2}} \\ &amp;-\frac{n}{2}\log2\pi
\end{aligned}
\]</span></p>
<p><span class="math display">\[
E(\boldsymbol{ \theta}) =
\color{yellow}{\frac{1}{2}\log\det{\mathbf{K}}} +
\color{cyan}{\frac{\mathbf{ y}^{\top}\mathbf{K}^{-1}\mathbf{ y}}{2}}
\]</span></p>
</section>
<section id="capacity-control-through-the-determinant"
class="slide level2">
<h2>Capacity Control through the Determinant</h2>
<p>The parameters are <em>inside</em> the covariance function (matrix).
<span class="math display">\[k_{i, j} = k(\mathbf{ x}_i, \mathbf{ x}_j;
\boldsymbol{ \theta})\]</span></p>
</section>
<section id="eigendecomposition-of-covariance" class="slide level2">
<h2>Eigendecomposition of Covariance</h2>
<p><span> <span class="math display">\[\mathbf{K}=
\mathbf{R}\boldsymbol{ \Lambda}^2 \mathbf{R}^\top\]</span></span></p>
<table>
<tr>
<td width="50%">
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimize-eigen.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
<td width="50%">
<span class="math inline">\(\boldsymbol{ \Lambda}\)</span> represents
distance on axes. <span class="math inline">\(\mathbf{R}\)</span> gives
rotation.
</td>
</tr>
</table>
</section>
<section id="eigendecomposition-of-covariance-1" class="slide level2">
<h2>Eigendecomposition of Covariance</h2>
<ul>
<li><span class="math inline">\(\boldsymbol{ \Lambda}\)</span> is
<em>diagonal</em>, <span
class="math inline">\(\mathbf{R}^\top\mathbf{R}=
\mathbf{I}\)</span>.</li>
<li>Useful representation since <span
class="math inline">\(\det{\mathbf{K}} = \det{\boldsymbol{ \Lambda}^2} =
\det{\boldsymbol{ \Lambda}}^2\)</span>.</li>
</ul>
</section>
<section id="capacity-control-coloryellowlog-detmathbfk"
class="slide level2">
<h2>Capacity control: <span class="math inline">\(\color{yellow}{\log
\det{\mathbf{K}}}\)</span></h2>
<script>
showDivs(0, 'gp-optimise-determinant');
</script>
<p><small></small>
<input id="range-gp-optimise-determinant" type="range" min="0" max="10" value="0" onchange="setDivs('gp-optimise-determinant')" oninput="setDivs('gp-optimise-determinant')">
<button onclick="plusDivs(-1, 'gp-optimise-determinant')">❮</button>
<button onclick="plusDivs(1, 'gp-optimise-determinant')">❯</button></p>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant000.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant001.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant002.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant003.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant004.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant005.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant006.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant007.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant008.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-determinant" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-determinant009.svg" width="80%" style=" ">
</object>
</div>
</section>
<section id="quadratic-data-fit" class="slide level2">
<h2>Quadratic Data Fit</h2>
</section>
<section id="data-fit-colorcyanfracmathbf-ytopmathbfk-1mathbf-y2"
class="slide level2">
<h2>Data Fit: <span class="math inline">\(\color{cyan}{\frac{\mathbf{
y}^\top\mathbf{K}^{-1}\mathbf{ y}}{2}}\)</span></h2>
<script>
showDivs(0, 'gp-optimise-quadratic');
</script>
<p><small></small>
<input id="range-gp-optimise-quadratic" type="range" min="0" max="2" value="0" onchange="setDivs('gp-optimise-quadratic')" oninput="setDivs('gp-optimise-quadratic')">
<button onclick="plusDivs(-1, 'gp-optimise-quadratic')">❮</button>
<button onclick="plusDivs(1, 'gp-optimise-quadratic')">❯</button></p>
<div class="gp-optimise-quadratic" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-quadratic000.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-quadratic" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-quadratic001.svg" width="80%" style=" ">
</object>
</div>
<div class="gp-optimise-quadratic" style="text-align:center;">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-quadratic002.svg" width="80%" style=" ">
</object>
</div>
<div class="figure">
<div id="gp-optimise-quadratic-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise-quadratic002.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
The data fit term of the Gaussian process is a quadratic loss centered
around zero. This has eliptical contours, the principal axes of which
are given by the covariance matrix.
</aside>
</section>
<section
id="eboldsymbol-theta-coloryellowfrac12logdetmathbfkcolorcyanfracmathbf-ytopmathbfk-1mathbf-y2"
class="slide level2">
<h2><span class="math display">\[E(\boldsymbol{ \theta}) =
\color{yellow}{\frac{1}{2}\log\det{\mathbf{K}}}+\color{cyan}{\frac{\mathbf{
y}^{\top}\mathbf{K}^{-1}\mathbf{ y}}{2}}\]</span></h2>
</section>
<section id="data-fit-term" class="slide level2">
<h2>Data Fit Term</h2>
<script>
showDivs(0, 'gp-optimise');
</script>
<p><small></small>
<input id="range-gp-optimise" type="range" min="0" max="10" value="0" onchange="setDivs('gp-optimise')" oninput="setDivs('gp-optimise')">
<button onclick="plusDivs(-1, 'gp-optimise')">❮</button>
<button onclick="plusDivs(1, 'gp-optimise')">❯</button></p>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise000.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise001.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise002.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise003.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise004.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise005.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise006.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise007.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise008.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise009.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise010.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise011.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise012.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise013.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise014.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise015.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise016.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise017.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise018.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise019.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
<div class="gp-optimise" style="text-align:center;">
<table>
<tr>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise020.svg" width style=" ">
</object>
</td>
<td width="50%">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gp-optimise021.svg" width style=" ">
</object>
</td>
</tr>
</table>
</div>
</section>
<section id="exponentiated-quadratic-covariance" class="slide level2">
<h2>Exponentiated Quadratic Covariance</h2>
<center>
<span class="math display">\[k(\mathbf{ x}, \mathbf{ x}^\prime) = \alpha
\exp\left(-\frac{\left\Vert \mathbf{ x}-\mathbf{ x}^\prime
\right\Vert_2^2}{2\ell^2}\right)\]</span>
</center>
<div class="figure">
<div id="eq-covariance-plot-figure" class="figure-frame">
<table>
<tr>
<td width="45%">
<object class data="../slides/diagrams/kern/eq_covariance.svg" width="100%" style=" ">
</object>
</td>
<td width="45%">
<div class="centered" style="">
<img class="negate" src="../slides/diagrams/kern/eq_covariance.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
The exponentiated quadratic covariance function.
</aside>
</section>
<section id="olympic-marathon-data" class="slide level2">
<h2>Olympic Marathon Data</h2>
<table>
<tr>
<td width="70%">
<ul>
<li>Gold medal times for Olympic Marathon since 1896.</li>
<li>Marathons before 1924 didn’t have a standardized distance.</li>
<li>Present results using pace per km.</li>
<li>In 1904 Marathon was badly organized leading to very slow
times.</li>
</ul>
</td>
<td width="30%">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//Stephen_Kiprotich.jpg" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
<small>Image from Wikimedia Commons <a href="http://bit.ly/16kMKHQ"
class="uri">http://bit.ly/16kMKHQ</a></small>
</td>
</tr>
</table>
</section>
<section id="olympic-marathon-data-1" class="slide level2">
<h2>Olympic Marathon Data</h2>
<div class="figure">
<div id="olympic-marathon-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//datasets/olympic-marathon.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Olympic marathon pace times since 1896.
</aside>
</section>
<section id="alan-turing" class="slide level2">
<h2>Alan Turing</h2>
<div class="figure">
<div id="turing-run-times-figure" class="figure-frame">
<table>
<tr>
<td width="50%">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//turing-times.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
<td width="50%">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//turing-run.jpg" width="50%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
Alan Turing, in 1946 he was only 11 minutes slower than the winner of
the 1948 games. Would he have won a hypothetical games held in 1946?
Source:
<a href="http://www.turing.org.uk/scrapbook/run.html" target="_blank">Alan
Turing Internet Scrapbook</a>.
</aside>
</section>
<section id="probability-winning-olympics" class="slide level2">
<h2>Probability Winning Olympics?</h2>
<ul>
<li>He was a formidable Marathon runner.</li>
<li>In 1946 he ran a time 2 hours 46 minutes.
<ul>
<li>That’s a pace of 3.95 min/km.</li>
</ul></li>
<li>What is the probability he would have won an Olympics if one had
been held in 1946?</li>
</ul>
</section>
<section id="gaussian-process-fit" class="slide level2">
<h2>Gaussian Process Fit</h2>
</section>
<section id="olympic-marathon-data-gp" class="slide level2">
<h2>Olympic Marathon Data GP</h2>
<div class="figure">
<div id="olympic-marathon-gp-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/olympic-marathon-gp.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Gaussian process fit to the Olympic Marathon data. The error bars are
too large, perhaps due to the outlier from 1904.
</aside>
</section>
<section id="della-gatta-gene-data" class="slide level2">
<h2>Della Gatta Gene Data</h2>
<ul>
<li>Given given expression levels in the form of a time series from
<span class="citation" data-cites="DellaGatta:direct08">Della Gatta et
al. (2008)</span>.</li>
</ul>
</section>
<section id="della-gatta-gene-data-1" class="slide level2">
<h2>Della Gatta Gene Data</h2>
<div class="figure">
<div id="della-gatta-gene-data-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//datasets/della-gatta-gene.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Gene expression levels over time for a gene from data provided by <span
class="citation" data-cites="DellaGatta:direct08">Della Gatta et al.
(2008)</span>. We would like to understand whether there is signal in
the data, or we are only observing noise.
</aside>
</section>
<section id="gene-expression-example" class="slide level2">
<h2>Gene Expression Example</h2>
<ul>
<li>Want to detect if a gene is expressed or not, fit a GP to each gene
<span class="citation" data-cites="Kalaitzis:simple11">Kalaitzis and
Lawrence (2011)</span>.</li>
</ul>
</section>
<section id="section-15" class="slide level2">
<h2></h2>
<div class="centered" style="">
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip1">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
Freddie Kalaitzis
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/freddie-kalaitzis.jpg" clip-path="url(#clip1)"/>
</svg>
</div>
<div class="figure">
<div id="a-simple-approach-to-ranking-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/1471-2105-12-180_1.png" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
The example is taken from the paper “A Simple Approach to Ranking
Differentially Expressed Gene Expression Time Courses through Gaussian
Process Regression.” <span class="citation"
data-cites="Kalaitzis:simple11">Kalaitzis and Lawrence (2011)</span>.
</aside>
<center>
<a href="http://www.biomedcentral.com/1471-2105/12/180"
class="uri">http://www.biomedcentral.com/1471-2105/12/180</a>
</center>
</section>
<section id="section-16" class="slide level2">
<h2></h2>
</section>
<section id="tp53-gene-data-gp" class="slide level2">
<h2>TP53 Gene Data GP</h2>
<div class="figure">
<div id="della-gatta-gene-gp-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/della-gatta-gene-gp.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Result of the fit of the Gaussian process model with the time scale
parameter initialized to 50 minutes.
</aside>
</section>
<section id="tp53-gene-data-gp-1" class="slide level2">
<h2>TP53 Gene Data GP</h2>
<div class="figure">
<div id="della-gatta-gene-gp2-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/della-gatta-gene-gp2.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Result of the fit of the Gaussian process model with the time scale
parameter initialized to 2000 minutes.
</aside>
</section>
<section id="tp53-gene-data-gp-2" class="slide level2">
<h2>TP53 Gene Data GP</h2>
<div class="figure">
<div id="della-gatta-gene-gp3-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/della-gatta-gene-gp3.svg" width="80%" style=" ">
</object>
</div>
</div>
<aside class="notes">
Result of the fit of the Gaussian process model with the noise
initialized low (standard deviation 0.1) and the time scale parameter
initialized to 20 minutes.
</aside>
</section>
<section id="multiple-optima" class="slide level2">
<h2>Multiple Optima</h2>
<div class="figure">
<div id="gp-multiple-optima000-figure" class="figure-frame">
<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/multiple-optima000.svg" width="50%" style=" ">
</object>
</div>
</div>
<aside class="notes">
</aside>
<!--
## Multiple Optima  {}



<object class="svgplot " data="https://mlatcl.github.io/mlai/./slides/diagrams//gp/multiple-optima001.svg" width="" style=" "></object>-->
</section>
<section id="example-prediction-of-malaria-incidence-in-uganda"
class="slide level2">
<h2>Example: Prediction of Malaria Incidence in Uganda</h2>
<div class="centered" style="">
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip2">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
Martin Mubangizi
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/martin-mubangizi.png" clip-path="url(#clip2)"/>
</svg>
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip3">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
Ricardo Andrade Pacecho
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/ricardo-andrade-pacheco.png" clip-path="url(#clip3)"/>
</svg>
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip4">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
John Quinn
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/john-quinn.jpg" clip-path="url(#clip4)"/>
</svg>
</div>
<ul>
<li>Work with Ricardo Andrade Pacheco, John Quinn and Martin Mubangizi
(Makerere University, Uganda)</li>
<li>See <a href="http://air.ug/research.html">AI-DEV Group</a>.</li>
<li>See <a href="https://diseaseoutbreaks.unglobalpulse.net/uganda/">UN
Global Pulse Disease Outbreaks Site</a></li>
</ul>
</section>
<section id="malaria-prediction-in-uganda" class="slide level2">
<h2>Malaria Prediction in Uganda</h2>
<div class="figure">
<div id="uganda-districts-2006-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/uganda-districts-2006.png" width="50%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Ugandan districts. Data SRTM/NASA from <a
href="https://dds.cr.usgs.gov/srtm/version2_1"
class="uri">https://dds.cr.usgs.gov/srtm/version2_1</a>.
</aside>
<div style="text-align:right">
<span class="citation"
data-cites="Andrade:consistent14 Mubangizi:malaria14">(Andrade-Pacheco
et al., 2014; Mubangizi et al., 2014)</span>
</div>
</section>
<section id="kapchorwa-district" class="slide level2">
<h2>Kapchorwa District</h2>
<div class="figure">
<div id="kapchorwa-district-in-uganda-figure" class="figure-frame">
<object class data="https://mlatcl.github.io/mlai/./slides/diagrams//health/Kapchorwa_District_in_Uganda.svg" width="50%" style=" ">
</object>
</div>
</div>
<aside class="notes">
The Kapchorwa District, home district of Stephen Kiprotich.
</aside>
</section>
<section id="tororo-district" class="slide level2">
<h2>Tororo District</h2>
<div class="figure">
<div id="tororo-district-in-uganda-figure" class="figure-frame">
<object class data="https://mlatcl.github.io/mlai/./slides/diagrams//health/Tororo_District_in_Uganda.svg" width="50%" style=" ">
</object>
</div>
</div>
<aside class="notes">
The Tororo district, where the sentinel site, Nagongera, is located.
</aside>
</section>
<section id="malaria-prediction-in-nagongera-sentinel-site"
class="slide level2">
<h2>Malaria Prediction in Nagongera (Sentinel Site)</h2>
<div class="figure">
<div id="sentinel-nagongera-figure" class="figure-frame">
<div class="centered" style="">
<img class="negate" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/sentinel_nagongera.png" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Sentinel and HMIS data along with rainfall and temperature for the
Nagongera sentinel station in the Tororo district.
</aside>
</section>
<section id="mubende-district" class="slide level2">
<h2>Mubende District</h2>
<div class="figure">
<div id="mubende-district-in-uganda-figure" class="figure-frame">
<object class data="https://mlatcl.github.io/mlai/./slides/diagrams//health/Mubende_District_in_Uganda.svg" width="50%" style=" ">
</object>
</div>
</div>
<aside class="notes">
The Mubende District.
</aside>
</section>
<section id="malaria-prediction-in-uganda-1" class="slide level2">
<h2>Malaria Prediction in Uganda</h2>
<div class="figure">
<div id="malaria-prediction-mubende-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/mubende.png" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Prediction of malaria incidence in Mubende.
</aside>
</section>
<section id="gp-school-at-makerere" class="slide level2">
<h2>GP School at Makerere</h2>
<div class="figure">
<div id="-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//gpss/1157497_513423392066576_1845599035_n.jpg" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
The project arose out of the Gaussian process summer school held at
Makerere in Kampala in 2013. The school led, in turn, to the Data
Science Africa initiative.
</aside>
</section>
<section id="kabarole-district" class="slide level2">
<h2>Kabarole District</h2>
<div class="figure">
<div id="kabarole-district-in-uganda-figure" class="figure-frame">
<object class data="https://mlatcl.github.io/mlai/./slides/diagrams//health/Kabarole_District_in_Uganda.svg" width="50%" style=" ">
</object>
</div>
</div>
<aside class="notes">
The Kabarole district in Uganda.
</aside>
</section>
<section id="early-warning-system" class="slide level2">
<h2>Early Warning System</h2>
<div class="figure">
<div id="kabarole-disease-over-time-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/kabarole.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
Estimate of the current disease situation in the Kabarole district over
time. Estimate is constructed with a Gaussian process with an additive
covariance funciton.
</aside>
</section>
<section id="early-warning-systems" class="slide level2">
<h2>Early Warning Systems</h2>
<div class="figure">
<div id="early-warning-system-map-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//health/monitor.gif" width="50%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
The map of Ugandan districts with an overview of the Malaria situation
in each district.
</aside>
</section>
<section id="additive-covariance" class="slide level2">
<h2>Additive Covariance</h2>
<center>
<span class="math display">\[k_f(\mathbf{ x}, \mathbf{ x}^\prime) =
k_g(\mathbf{ x}, \mathbf{ x}^\prime) + k_h(\mathbf{ x}, \mathbf{
x}^\prime)\]</span>
</center>
<div class="figure">
<div id="add-covariance-plot-figure" class="figure-frame">
<table>
<tr>
<td width="45%">
<object class data="../slides/diagrams/kern/add_covariance.svg" width="100%" style=" ">
</object>
</td>
<td width="45%">
<div class="centered" style="">
<img class="negate" src="../slides/diagrams/kern/add_covariance.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
An additive covariance function formed by combining a linear and an
exponentiated quadratic covariance functions.
</aside>
</section>
<section id="section-17" class="slide level2">
<h2></h2>
<div class="centered" style="">
<svg viewBox="0 0 200 200" style="width:15%">
<defs> <clipPath id="clip5">
<style>
circle {
  fill: black;
}
</style>
<circle cx="100" cy="100" r="100"/> </clipPath> </defs>
<title>
Aki Vehtari
</title>
<image preserveAspectRatio="xMinYMin slice" width="100%" xlink:href="https://mlatcl.github.io/mlai/./slides/diagrams//people/aki-vehtari.jpg" clip-path="url(#clip5)"/>
</svg>
</div>
<div class="figure">
<div id="bialik-friday-the-13th-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//ml/bialik-fridaythe13th-1.png" width="70%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
This is a retrospective analysis of US births by Aki Vehtari. The
challenges of forecasting. Even with seasonal and weekly effects removed
there are significant effects on holidays, weekends, etc.
</aside>
</section>
<section id="gelman-book" class="slide level2">
<h2>Gelman Book</h2>
<div class="figure">
<div id="bayesian-data-analysis-figure" class="figure-frame">
<table>
<tr>
<td width="50%">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//ml/bda_cover_1.png" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
<td width="50%">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//ml/bda_cover.png" width="80%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
Two different editions of Bayesian Data Analysis <span class="citation"
data-cites="Gelman:bayesian13">(Gelman et al., 2013)</span>.
</aside>
<div style="text-align:right">
<span class="citation" data-cites="Gelman:bayesian13">Gelman et al.
(2013)</span>
</div>
</section>
<section id="basis-function-covariance" class="slide level2">
<h2>Basis Function Covariance</h2>
<center>
<span class="math display">\[k(\mathbf{ x}, \mathbf{ x}^\prime) =
\boldsymbol{ \phi}(\mathbf{ x})^\top \boldsymbol{ \phi}(\mathbf{
x}^\prime)\]</span>
</center>
<div class="figure">
<div id="basis-covariance-plot-figure" class="figure-frame">
<table>
<tr>
<td width="45%">
<object class data="../slides/diagrams/kern/basis_covariance.svg" width="100%" style=" ">
</object>
</td>
<td width="45%">
<div class="centered" style="">
<img class="negate" src="../slides/diagrams/kern/basis_covariance.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
A covariance function based on a non-linear basis given by <span
class="math inline">\(\boldsymbol{ \phi}(\mathbf{ x})\)</span>.
</aside>
</section>
<section id="brownian-covariance" class="slide level2">
<h2>Brownian Covariance</h2>
<center>
<span class="math display">\[k(t, t^\prime)=\alpha \min(t,
t^\prime)\]</span>
</center>
<div class="figure">
<div id="brownian-covariance-plot-figure" class="figure-frame">
<table>
<tr>
<td width="45%">
<object class data="../slides/diagrams/kern/brownian_covariance.svg" width="100%" style=" ">
</object>
</td>
<td width="45%">
<div class="centered" style="">
<img class="negate" src="../slides/diagrams/kern/brownian_covariance.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
Brownian motion covariance function.
</aside>
</section>
<section id="mlp-covariance" class="slide level2">
<h2>MLP Covariance</h2>
<center>
<span class="math display">\[k(\mathbf{ x}, \mathbf{ x}^\prime) = \alpha
\arcsin\left(\frac{w \mathbf{ x}^\top \mathbf{ x}^\prime +
b}{\sqrt{\left(w \mathbf{ x}^\top \mathbf{ x}+ b + 1\right)\left(w
\left.\mathbf{ x}^\prime\right.^\top \mathbf{ x}^\prime + b +
1\right)}}\right)\]</span>
</center>
<div class="figure">
<div id="mlp-covariance-plot-figure" class="figure-frame">
<table>
<tr>
<td width="45%">
<object class data="../slides/diagrams/kern/mlp_covariance.svg" width="100%" style=" ">
</object>
</td>
<td width="45%">
<div class="centered" style="">
<img class="negate" src="../slides/diagrams/kern/mlp_covariance.gif" width="100%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</td>
</tr>
</table>
</div>
</div>
<aside class="notes">
The multi-layer perceptron covariance function. This is derived by
considering the infinite limit of a neural network with probit
activation functions.
</aside>
</section>
<section id="gpss-gaussian-process-summer-school" class="slide level2">
<h2>GPSS: Gaussian Process Summer School</h2>
<table>
<tr>
<td width="60%">
<ul>
<li><a href="http://gpss.cc" class="uri">http://gpss.cc</a></li>
<li>Next one is in Sheffield in <em>September 2019</em>.</li>
<li>Many lectures from past meetings available online</li>
</ul>
</td>
<td width="40%">
<div style="width:1.5cm;text-align:center">

</div>
</td>
</tr>
</table>
</section>
<section id="gpy-a-gaussian-process-framework-in-python"
class="slide level2">
<h2>GPy: A Gaussian Process Framework in Python</h2>
<div class="figure">
<div id="gpy-software-figure" class="figure-frame">
<div class="centered" style="">
<img class="" src="https://mlatcl.github.io/mlai/./slides/diagrams//gp/gpy.png" width="70%" height="auto" align="center" style="background:none; border:none; box-shadow:none; display:block; margin-left:auto; margin-right:auto;vertical-align:middle">
</div>
</div>
</div>
<aside class="notes">
GPy is a BSD licensed software code base for implementing Gaussian
process models in Python. It is designed for teaching and modelling. We
welcome contributions which can be made through the GitHub repository <a
href="https://github.com/SheffieldML/GPy"
class="uri">https://github.com/SheffieldML/GPy</a>
</aside>
<center>
<a href="https://github.com/SheffieldML/GPy"
class="uri">https://github.com/SheffieldML/GPy</a>
</center>
</section>
<section id="gpy-a-gaussian-process-framework-in-python-1"
class="slide level2">
<h2>GPy: A Gaussian Process Framework in Python</h2>
<ul>
<li>BSD Licensed software base.</li>
<li>Wide availability of libraries, ‘modern’ scripting language.</li>
<li>Allows us to set projects to undergraduates in Comp Sci that use
GPs.</li>
<li>Available through GitHub <a
href="https://github.com/SheffieldML/GPy"
class="uri">https://github.com/SheffieldML/GPy</a></li>
<li>Reproducible Research with Jupyter Notebook.</li>
</ul>
</section>
<section id="features" class="slide level2">
<h2>Features</h2>
<ul>
<li>Probabilistic-style programming (specify the model, not the
algorithm).</li>
<li>Non-Gaussian likelihoods.</li>
<li>Multivariate outputs.</li>
<li>Dimensionality reduction.</li>
<li>Approximations for large data sets.</li>
</ul>
</section>
<section id="thanks" class="slide level2 scrollable">
<h2 class="scrollable">Thanks!</h2>
<ul>
<li><p>twitter: <a
href="https://twitter.com/lawrennd">@lawrennd</a></p></li>
<li><p>podcast: <a href="http://thetalkingmachines.com">The Talking
Machines</a></p></li>
<li><p>newspaper: <a
href="http://www.theguardian.com/profile/neil-lawrence">Guardian Profile
Page</a></p></li>
<li><p>blog posts:</p>
<p><a
href="http://inverseprobability.com/2014/07/01/open-data-science">Open
Data Science</a></p></li>
</ul>
</section>
<section id="references" class="slide level2 unnumbered scrollable">
<h2 class="unnumbered scrollable">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent"
role="list">
<div id="ref-Andrade:consistent14" class="csl-entry" role="listitem">
Andrade-Pacheco, R., Mubangizi, M., Quinn, J., Lawrence, N.D., 2014.
Consistent mapping of government malaria records across a changing
territory delimitation. Malaria Journal 13. <a
href="https://doi.org/10.1186/1475-2875-13-S1-P5">https://doi.org/10.1186/1475-2875-13-S1-P5</a>
</div>
<div id="ref-DellaGatta:direct08" class="csl-entry" role="listitem">
Della Gatta, G., Bansal, M., Ambesi-Impiombato, A., Antonini, D.,
Missero, C., Bernardo, D. di, 2008. Direct targets of the TRP63
transcription factor revealed by a combination of gene expression
profiling and reverse engineering. Genome Research 18, 939–948. <a
href="https://doi.org/10.1101/gr.073601.107">https://doi.org/10.1101/gr.073601.107</a>
</div>
<div id="ref-Gelman:bayesian13" class="csl-entry" role="listitem">
Gelman, A., Carlin, J.B., Stern, H.S., Dunson, D.B., Vehtari, A., Rubin,
D.B., 2013. Bayesian data analysis, 3rd ed. Chapman; Hall.
</div>
<div id="ref-Kalaitzis:simple11" class="csl-entry" role="listitem">
Kalaitzis, A.A., Lawrence, N.D., 2011. A simple approach to ranking
differentially expressed gene expression time courses through
<span>Gaussian</span> process regression. BMC Bioinformatics 12. <a
href="https://doi.org/10.1186/1471-2105-12-180">https://doi.org/10.1186/1471-2105-12-180</a>
</div>
<div id="ref-MacKay:bayesian92" class="csl-entry" role="listitem">
MacKay, D.J.C., 1992. Bayesian methods for adaptive models (PhD thesis).
California Institute of Technology.
</div>
<div id="ref-Mubangizi:malaria14" class="csl-entry" role="listitem">
Mubangizi, M., Andrade-Pacheco, R., Smith, M.T., Quinn, J., Lawrence,
N.D., 2014. Malaria surveillance with multiple data sources using
<span>Gaussian</span> process models, in: 1st International Conference
on the Use of Mobile <span>ICT</span> in Africa.
</div>
<div id="ref-Neal:bayesian94" class="csl-entry" role="listitem">
Neal, R.M., 1994. Bayesian learning for neural networks (PhD thesis).
Dept. of Computer Science, University of Toronto.
</div>
</div>
</section>
    </div>
  </div>

  <script src="https://unpkg.com/reveal.js@3.9.2/lib/js/head.min.js"></script>
  <script src="https://unpkg.com/reveal.js@3.9.2/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display controls in the bottom right corner
        controls: true,
        // Display a presentation progress bar
        progress: true,
        // Push each slide change to the browser history
        history: true,
        // Enable keyboard shortcuts for navigation
        keyboard: true,
        // Enable the slide overview mode
        overview: true,
        // Vertical centering of slides
        center: true,
        // Enables touch navigation on devices with touch input
        touch: true,
        // Turns fragments on and off globally
        fragments: true,
        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,
        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,
        // Stop auto-sliding after user input
        autoSlideStoppable: true,
        // Transition style
        transition: 'slide', // none/fade/slide/convex/concave/zoom
        // Transition speed
        transitionSpeed: 'default', // default/fast/slow
        // Transition style for full page slide backgrounds
        backgroundTransition: 'fade', // none/fade/slide/convex/concave/zoom
        // Number of slides away from the current that are visible
        viewDistance: 3,
        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // Optional reveal.js plugins
        dependencies: [
          { src: 'https://unpkg.com/reveal.js@3.9.2/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'https://unpkg.com/reveal.js@3.9.2/plugin/zoom-js/zoom.js', async: true },
          { src: 'https://unpkg.com/reveal.js@3.9.2/plugin/math/math.js', async: true },
          { src: 'https://unpkg.com/reveal.js@3.9.2/plugin/notes/notes.js', async: true }
        ]
      });
    </script>
    </body>
</html>
